tests/arch/arm/arm_custom_interrupt/src/arm_custom_interrupt.c

```diff
@@ -0,0 +1,210 @@
+/*
+ * Copyright Meta Platforms, Inc. and its affiliates.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/ztest.h>
+#include <zephyr/arch/cpu.h>
+#include <cmsis_core.h>
+#include <zephyr/sys/barrier.h>
+
+unsigned int sw_irq_number = (unsigned int)(-1);
+static volatile bool custom_init_called;
+static volatile bool custom_enable_called;
+static volatile bool custom_disable_called;
+static volatile bool custom_set_priority_called;
+static volatile bool custom_eoi_called;
+static volatile bool irq_handler_called;
+
+/* Define out custom SoC interrupt controller interface methods.
+ * These closely match the normal Cortex-M implementations.
+ */
+
+#define NUM_IRQS_PER_REG  32
+#define REG_FROM_IRQ(irq) (irq / NUM_IRQS_PER_REG)
+#define BIT_FROM_IRQ(irq) (irq % NUM_IRQS_PER_REG)
+
+void z_soc_irq_init(void)
+{
+	int irq = 0;
+
+	for (; irq < CONFIG_NUM_IRQS; irq++) {
+		NVIC_SetPriority((IRQn_Type)irq, _IRQ_PRIO_OFFSET);
+	}
+
+	custom_init_called = true;
+}
+
+void z_soc_irq_enable(unsigned int irq)
+{
+	if (irq == sw_irq_number) {
+		custom_enable_called = true;
+	}
+	NVIC_EnableIRQ((IRQn_Type)irq);
+}
+
+void z_soc_irq_disable(unsigned int irq)
+{
+	if (irq == sw_irq_number) {
+		custom_disable_called = true;
+	}
+	NVIC_DisableIRQ((IRQn_Type)irq);
+}
+
+int z_soc_irq_is_enabled(unsigned int irq)
+{
+	return NVIC->ISER[REG_FROM_IRQ(irq)] & BIT(BIT_FROM_IRQ(irq));
+}
+
+void z_soc_irq_eoi(unsigned int irq)
+{
+	if (irq == sw_irq_number) {
+		custom_eoi_called = true;
+	}
+}
+
+__attribute__((naked)) unsigned int z_soc_irq_get_active(void)
+{
+	__asm("mrs r0, IPSR\n"
+	      "bx    lr\n");
+}
```

Comment:
The only reason I did it this way was to implement it as a naked function to make as fast as possible. We could use ```__get_IPSR()``` but then we would have to include the whole C preamble since it's not implemented in a manner that guarantees registers are preserved. It's test code and not a hill worth dying on though, so if you feel this is clearer I'm happy to change.

Reply:
I’m trying to limit appearance of ASM where it isn’t strictly necessary.

[`__get_IPSR()`](https://github.com/zephyrproject-rtos/cmsis/blob/master/CMSIS/Core/Include/cmsis_gcc.h#L1031) is a static inline that lets the compiler choose which register to use.

So there typically won’t be any preamble for inlining. Your function could be marked [`always_inline`](https://gcc.gnu.org/onlinedocs/gcc/Common-Function-Attributes.html#index-always_005finline-function-attribute) too to make it transparent too.

Reply:
You are right, this is now done.